compiler/plc_diagnostics/src/diagnostics.rs

```diff
@@ -1,855 +1,482 @@
+    /// Seconday locations relevant to the diagnostics
+    secondary_locations: Option<Vec<SourceLocation>>,
+    /// Severity of the error being reported
+    severity: Severity,
+    /// Error code for reference in the documentation
+    error_code: &'static str,
+    /// Children of the current diagnostic
+    sub_diagnostics: Vec<Diagnostic>,
+    /// If the diagnostic is caused by an error, this field contains the original error
+    internal_error: Option<anyhow::Error>,
 }
 
-impl Display for Diagnostic {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}: {}", self.get_type(), self.get_message())?;
-        let location = self.get_location();
-        if !location.is_undefined() {
-            write!(f, " at: {location}")
-        } else {
-            Ok(())
-        }
+impl std::error::Error for Diagnostic {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        self.internal_error.as_ref().and_then(|it| it.source())
     }
 }
 
-impl<T: Error> From<T> for Diagnostic {
-    fn from(e: T) -> Self {
-        Diagnostic::GeneralError { message: e.to_string(), err_no: ErrNo::general__io_err }
+impl From<std::io::Error> for Diagnostic {
+    fn from(value: std::io::Error) -> Self {
+        Diagnostic::error(value.to_string()).with_error_code("E002").with_internal_error(value.into())
     }
 }
 
+/// Builder for Diagnostics
 impl Diagnostic {
-    /// Creates a new diagnostic with additional ranges
-    pub fn with_extra_ranges(self, ranges: &[SourceLocation]) -> Diagnostic {
-        match self {
-            Diagnostic::SyntaxError { message, mut range, err_no } => {
-                range.extend_from_slice(ranges);
-                Diagnostic::SyntaxError { message, range, err_no }
-            }
-            Diagnostic::SemanticError { message, mut range, err_no } => {
-                range.extend_from_slice(ranges);
-                Diagnostic::SyntaxError { message, range, err_no }
-            }
-            Diagnostic::ImprovementSuggestion { message, mut range } => {
-                range.extend_from_slice(ranges);
-                Diagnostic::ImprovementSuggestion { message, range }
-            }
-            _ => self,
+    pub fn new(message: impl Into<String>, severity: Severity) -> Self {
+        Diagnostic {
+            message: message.into(),
+            severity,
+            primary_location: SourceLocation::undefined(),
+            secondary_locations: Default::default(),
+            error_code: Default::default(),
+            sub_diagnostics: Default::default(),
+            internal_error: Default::default(),
         }
     }
 
-    pub fn syntax_error(message: &str, range: SourceLocation) -> Diagnostic {
-        Diagnostic::SyntaxError {
-            message: message.to_string(),
-            range: vec![range],
-            err_no: ErrNo::syntax__generic_error,
-        }
+    pub fn error(message: impl Into<String>) -> Self {
```

Comment:
Yes, I am still not happy with this, but it would be the next step. I am thinking of a macro that also includes the E number and can validate (Not sure how it will look because i was hoping not to limit the E numbers to this crate)
The "With Error Code" idea is to keep it fluent. I'm still open to changes here but I would say we first use the default E001 and do this in the next step